api-conversion/src/test/java/io/strimzi/kafka/api/conversion/cli/FullUpgradeLifecycleIT.java

```diff
@@ -0,0 +1,205 @@
+            assertThat(actualKafka1.getSpec().getKafka().getListeners().getKafkaListeners(), is(nullValue()));
+            assertThat(actualKafka1.getSpec().getKafka().getListeners().getGenericKafkaListeners(), is(notNullValue()));
+            assertThat(actualKafka1.getSpec().getKafka().getListeners().getGenericKafkaListeners().size(), is(3));
+        } finally {
+            opV1Beta2.inNamespace(NAMESPACE).withName("kafka1").delete();
+            CliTestUtils.deleteV1Crds(CLUSTER);
+        }
+    }
+
+    @Test
+    public void testFullLifecycleOnSecondTry() {
+        MixedOperation<Kafka, KafkaList, Resource<Kafka>> opV1Beta2 = Crds.kafkaOperation(client, ApiVersion.V1BETA2.toString());
+        MixedOperation<Kafka, KafkaList, Resource<Kafka>> opV1Beta1 = Crds.kafkaOperation(client, ApiVersion.V1BETA1.toString());
+
+        try {
+            Kafka kafka1 = new KafkaBuilder()
+                    .withNewMetadata()
+                        .withName("kafka1")
+                    .endMetadata()
+                    .withNewSpec()
+                        .withNewZookeeper()
+                            .withReplicas(3)
+                            .withNewEphemeralStorage()
+                            .endEphemeralStorage()
+                        .endZookeeper()
+                        .withNewKafka()
+                            .withVersion("2.7.0")
+                            .withReplicas(3)
+                            .withNewListeners()
+                                .withKafkaListeners(new KafkaListenersBuilder()
+                                        .withNewPlain()
+                                        .endPlain()
+                                        .withNewTls()
+                                        .endTls()
+                                        .withNewKafkaListenerExternalLoadBalancer()
+                                        .endKafkaListenerExternalLoadBalancer()
+                                        .build())
+                            .endListeners()
+                            .withNewEphemeralStorage()
+                            .endEphemeralStorage()
+                        .endKafka()
+                        .withNewEntityOperator()
+                            .withNewUserOperator()
+                            .endUserOperator()
+                            .withNewTopicOperator()
+                            .endTopicOperator()
+                        .endEntityOperator()
+                    .endSpec()
+                    .build();
+
+            opV1Beta1.inNamespace(NAMESPACE).create(kafka1);
+
+            // First try with unconverted resources => should fail
+            CommandLine cmd = new CommandLine(new EntryCommand());
+            StringWriter sw = new StringWriter();
+            PrintWriter pw = new PrintWriter(sw);
+            cmd.setOut(pw);
+            cmd.setErr(pw);
+
+            int exitCode = cmd.execute("crd-upgrade");
```

Comment:
I assume this fails because the tool does some check on the resources and rejects the upgrade, right?

Reply:
The tool will get the resource from Kube API and call replace on it. Since that is using the `v1beta2` API, it will reject any incompatible resources. We need to do this replace anyway to have the resource stored under the new version => so this serves for both - double check the validity and update the stored version. (The conversion alone should convert the API, but it would still be stored as v1beta1 in Etcd because that is the stored version at that time).